### PR TITLE
feat: new sync option to use ignore diff configs during sync

### LINKED ISSUE #2913: Maintain difference in cluster and git values for specific fields
# Summary
If another controller modifies a field in a resource and that field is required by the resource to be created, then that resource cannot be compatible with GitOps. The manifest stored in Git requires that field in order to be created, but another controller will modify that field once created making it out of sync. Argo CD should have an annotation called `argocd.argoproj.io/respect-existing-fields` that accepts JSON paths to fields that are managed by other controllers. That annotation should use stored in the Kubernetes instead of applying what is in Git when Argo CD does a diff or apply. This is a step further than https://argoproj.github.io/argo-cd/user-guide/diffing/#application-level-configuration

# Motivation
Istio Virtual Services require a weight when two destinations are listed in a route. If a user creates the Virtual Service using GitOps and another controller modifies those weights to shape traffic, the Virtual Service will be out of sync with what is stored in Git. If there is a Sync, the Virtual Service will be revert to the Git state and then the other controller should revert it back.

# Proposal

How do you think this should be implemented?

The application controller should look for this annotation on all objects within the repo while executing a diff or sync. If found, the diffing logic should not figure this information into Diff status, and the apply logic should not change this value in the cluster.

Questions/Concerns that should be addressed during the implementation:
- How should the annotation be applied if the annotation itself is modified in Git?
- The `argocd app get manifest` command should return the actual Git contents instead of the modified Git contents.
- The annotation should be able to index an entire array (i.e. `spec.http.[*].route[*].weight` means check each entry of http and the entries within each route.)
- Allow multiple fields to ignore

### PR DESCRIPTION
**Pull request by leoluz**
This PR introduces a new sync option to use ignore diff configs during sync.

Fix https://github.com/argoproj/argo-cd/issues/2913

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

File: cmd/argocd/commands/admin/settings.go [hunk 1]
```diff
@@ -406,6 +406,7 @@ argocd admin settings resource-overrides ignore-differences ./deploy.yaml --argo
 					return
 				}
 
+				// TODO validate if this IgnoreNormalizer needs to ignore managed fields as well.
```
Comment (alexmt) [inline cmd/argocd/commands/admin/settings.go:409]: I think there is no need to ignore managed fields here. The `argocd admin settings resource-overrides ignore-differences` command is supposed to help users to test fields configured in `ignoreDifferences.jsonPointers` or `ignoreDifferences.jqPathExpressions`. The command accepts sample resource manifest, removes ignored fields and prints the difference to visualize which fields were ignored.

In order to test `managedFieldsManagers` user would have to provide sample git and sample live k8s resource, so we would need a separate command. But `managedFieldsManagers` is much easier to configure and there is no need so such command at all.

File: util/argo/diff/normalize.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-package argo
+package diff
```
Comment (alexmt) [inline util/argo/diff/normalize.go:1]: Thanks for moving diff related logic into sub package ! 👍

**Review by alexmt — CHANGES_REQUESTED** [review]
Awesome PR. I have nothing to add regarding the implementation.

The only missing thing is e2e test. Can you please add e2e test that verifies the new setting? You can find similar test here:

https://github.com/argoproj/argo-cd/blob/5da1522b17bda3bd005595606a20b965cd3987d4/test/e2e/app_management_test.go#L515

File: cmd/argocd/commands/admin/settings.go (hunk 1 above)
Comment (leoluz) [reply cmd/argocd/commands/admin/settings.go:409]: Makes sense. Thanks for the explanation. Added proper comment in the code.

**Comment by leoluz** [comment]
@alexmt e2e test implemented

File: test/e2e/app_management_test.go [hunk 3]
```diff
@@ -547,6 +548,60 @@ func TestResourceDiffing(t *testing.T) {
...
+				JSONPointers:          []string{"/spec/template/spec/containers/0/image"},
+			},
+		}}).
+		When().
+		Refresh(RefreshTypeNormal).
+		Then().
+		Expect(SyncStatusIs(SyncStatusCodeSynced)).
+		Given().
+		When().
+		Sync().
```
Comment (alexmt) [inline test/e2e/app_management_test.go:575]: I think app should be patched before syncing, right? Can you swap these two lines?

**Review by alexmt — APPROVED** [review]
LGTM except minor comment in test. Merging once it is resolved.

File: test/e2e/app_management_test.go (hunk 3 above)
Comment (leoluz) [reply test/e2e/app_management_test.go:575]: Actually this is to test the old behaviour before enabling the `RespectIgnoreDifferences=true`.

**Pull request merged**